src/lib/operators/delete.cpp

```diff
@@ -13,127 +13,144 @@
+
+  // If there is nothing to delete, early out. Code below expects there to be Chunks in values_to_delete
+  if (values_to_delete->chunk_count() == 0) {
+    return nullptr;
+  }
+
+  // The first segment of the first Chunk of values_to_delete determines the referenced_table all other
+  // ReferenceSegments are expected to reference
+  const auto upper_left_reference_segment = std::dynamic_pointer_cast<const ReferenceSegment>(
+      values_to_delete->get_chunk(ChunkID{0})->get_segment(ColumnID{0}));
+  const auto referenced_table = upper_left_reference_segment->referenced_table();
 
   context->register_read_write_operator(std::static_pointer_cast<AbstractReadWriteOperator>(shared_from_this()));
 
-  _table = StorageManager::get().get_table(_table_name);
+  _stored_table = StorageManager::get().get_table(_table_name);
   _transaction_id = context->transaction_id();
 
-  const auto values_to_delete = input_table_left();
+  // If values_to_delete references a pruned table (and not the stored table itself), build a mapping from ChunkIDs in
+  // the pruned table (`referenced_table`) to the actual ChunkIDs in the stored_table
+  auto stored_chunk_id_by_referenced_chunk_id = std::optional<std::vector<ChunkID>>{};
+  if (_stored_table != referenced_table) {
+    auto stored_chunk_id_by_chunk = std::unordered_map<std::shared_ptr<const Chunk>, ChunkID>{};
+    for (auto stored_chunk_id = ChunkID{0}; stored_chunk_id < _stored_table->chunk_count(); ++stored_chunk_id) {
+      stored_chunk_id_by_chunk.emplace(_stored_table->get_chunk(stored_chunk_id), stored_chunk_id);
+    }
+
+    stored_chunk_id_by_referenced_chunk_id.emplace(referenced_table->chunk_count());
+    for (auto referenced_chunk_id = ChunkID{0}; referenced_chunk_id < referenced_table->chunk_count();
+         ++referenced_chunk_id) {
+      const auto stored_chunk_id = stored_chunk_id_by_chunk.at(referenced_table->get_chunk(referenced_chunk_id));
+      (*stored_chunk_id_by_referenced_chunk_id)[referenced_chunk_id] = stored_chunk_id;
+    }
+  }
 
   for (ChunkID chunk_id{0}; chunk_id < values_to_delete->chunk_count(); ++chunk_id) {
     const auto chunk = values_to_delete->get_chunk(chunk_id);
 
-    // we have already verified that all segments reference the same table
+    DebugAssert(
+        std::dynamic_pointer_cast<const ReferenceSegment>(chunk->get_segment(ColumnID{0}))->referenced_table() ==
+            referenced_table,
+        "All segments in values_to_delete must reference the same table");
+    DebugAssert(chunk->references_exactly_one_table(),
+                "All segments in values_to_delete must reference the same table");
+
     const auto first_segment = std::static_pointer_cast<const ReferenceSegment>(chunk->get_segment(ColumnID{0}));
     const auto pos_list = first_segment->pos_list();
 
-    _pos_lists.emplace_back(pos_list);
+    DebugAssert(std::all_of(chunk->segments().begin(), chunk->segments().end(),
```

Comment:
This is silly, tbh. values_to_delete can contain multiple columns, but we just use the leftmost. To avoid confusion, we check the all other columns have the same poslist as the leftmost column...

Reply:
This is to avoid cases like `DELETE FROM t t1, t t2 WHERE t1.a = 3` where it is unclear what is deleted.

Reply:
We might end up with cases where all reference segments point to the same table but to different rows. I am not sure how we deal with that right now, but as an extra safety, this debug assert is there.

Reply:
(What I meant with "silly" is just that the Delete has to deal with the input table containing multiple columns in the first place, despite it never using anything but the first column - we need a multi-line assert to ensure the validity of data (columns 1->n) that we never use anyway - that's "silly", imo)

Reply:
> despite it never using anything but the first column

I would rather argue that it uses *all* columns.

Do you have a different suggestion though?

Reply:
Na, just made the original comment to ensure the reader that I placed an awkward multiline assert intentionally.